### PR TITLE
Addition of single iteration Patatrack and LST HLT configurations and workflows

### PR DESCRIPTION
**Pull request by VourMa**
### PR description

This is a procModifier-based implementation of LST in HLT. It reuses the `trackingLST` procModifier to enable LST track building, and it introduces a new one, `seedingLST`, which is not too useful for the HLT currently but it may become useful also for the offline setup in the near future.

Another modifier, `singleIterPatatrack` is introduced and it allows HLT to run only one iteration of tracking based on Patatrack pixel seeds. It works both with CKF and with LST building (when combined with the procModifier above).

Three new variations of the HLT timing workflow are also added (single iteration, single iteration LST, LST).

The most important results have been summarized in [this presentation](https://indico.cern.ch/event/1482695/#7-lst-in-phase-2-hlt-configura).

<details><summary>Instructions on how to run</summary>

```sh
cmsrel CMSSW_14_2_0_pre4
cd CMSSW_14_2_0_pre4/src
cmsenv
git cms-init
git cms-addpkg Configuration HLTrigger
git remote add SegLink git@github.com:SegmentLinking/cmssw.git
git fetch SegLink
git checkout CMSSW_14_2_0_pre4_singleIter_LST_HLT
scram b -j 16

# Rerun L1, needed because the HLT samples I am using are produced with 14_0_X
cmsDriver.py Phase2 -s L1,L1TrackTrigger \
--conditions auto:phase2_realistic_T33 \
--geometry ExtendedRun4D110 \
--era Phase2C17I13M9 \
--eventcontent FEVTDEBUGHLT \
--datatier GEN-SIM-DIGI-RAW-MINIAOD \
--customise SLHCUpgradeSimulations/Configuration/aging.customise_aging_1000,Configuration/DataProcessing/Utils.addMonitoring,L1Trigger/Configuration/customisePhase2FEVTDEBUGHLT.customisePhase2FEVTDEBUGHLT,L1Trigger/Configuration/customisePhase2TTOn110.customisePhase2TTOn110 \
--filein /store/mc/Phase2Spring24DIGIRECOMiniAOD/TT_TuneCP5_14TeV-powheg-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200_AllTP_140X_mcRun4_realistic_v4-v1/2560000/11d1f6f0-5f03-421e-90c7-b5815197fc85.root \
--fileout file:output_Phase2_L1T.root \
--inputCommands="keep *, drop l1tPFJets_*_*_*, drop l1tTrackerMuons_l1tTkMuonsGmt*_*_HLT" \
--outputCommands="drop l1tTrackerMuons_l1tTkMuonsGmt*_*_HLT" \
--mc -n 100 --nThreads 10

# Rerun L1 emulator + HLT (needed for any sample)
# In terms of procModifiers:
#    None -> Legacy seeding, CKF building
#    alpaka -> Patatrack seeding
#    singleIterPatatrack -> Single iteration tracking
#    trackingLST -> LST building for the initialStep
#    seedingLST -> LST seeding for the hightPtTripletStep
cmsDriver.py Phase2 -s L1P2GT,HLT:75e33 --processName=HLTX \
--conditions auto:phase2_realistic_T33 \
--geometry ExtendedRun4D110 \
--era Phase2C17I13M9 \
--eventcontent FEVTDEBUGHLT \
--customise SLHCUpgradeSimulations/Configuration/aging.customise_aging_1000 \
--filein file:output_Phase2_L1T.root \
--fileout file:Phase2_L1P2GT_HLT.root \
--inputCommands="keep *, drop *_hlt*_*_HLT, drop triggerTriggerFilterObjectWithRefs_l1t*_*_HLT" \
--outputCommands='keep *_*InitialStepTrack*LST*_*_*, keep *_*HighPtTripletStepTrack*_*_*' \
-n 100 --nThreads 10 \
--procModifiers alpaka,trackingLST,seedingLST

cmsDriver.py DQM -s VALIDATION:hltMultiTrackValidation \
--conditions auto:phase2_realistic_T33 \
--geometry ExtendedRun4D110 \
--era Phase2C17I13M9 \
--datatier DQMIO \
--eventcontent DQM \
--filein file:Phase2_L1P2GT_HLT.root \
--hltProcess HLTX \
--fileout DQM.root \
-n 100 --nThreads 10

cmsDriver.py HARVEST -s HARVESTING:@trackingOnlyValidation+@trackingOnlyDQM+postProcessorHLTtrackingSequence \
--filein file:DQM.root \
--scenario pp \
--filetype DQM \
--conditions auto:phase2_realistic_T33 \
--mc -n 100

mv DQM_V0001_R000000001__Global__CMSSW_X_Y_Z__RECO.root configName.root
makeTrackValidationPlots.py configName.root --extended
```

</details>

FYI @slava77 @rovere @lguzzi

### CODE REVIEW

**Comment by VourMa** [comment]
I am not sure whether I should use the procModifiers introduced in this PR always with the `alpaka` procModifier. Feel free to give feedback on that and I can modify the necessary lines.

**Comment by mmusich** [comment]
test parameters:
   * relvals_opt = --what upgrade
   * workflows = 29634.753, 29634.754, 29634.755
   * enable = hlt_p2_integration, hlt_p2_timing

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltHighPtTripletStepSeedTracksLST_cfi.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import FWCore.ParameterSet.Config as cms
+
+hltHighPtTripletStepSeedTracksLST = cms.EDProducer(
+    "TrackFromSeedProducer",
+    src = cms.InputTag("hltHighPtTripletStepSeeds"),
+    beamSpot = cms.InputTag("offlineBeamSpot"),
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltHighPtTripletStepSeedTracksLST_cfi.py:6]: is this correct? shouldn't it be `hltOnlineBeamSpot` ?

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepSeedTracksLST_cfi.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import FWCore.ParameterSet.Config as cms
+
+hltInitialStepSeedTracksLST = cms.EDProducer(
+    "TrackFromSeedProducer",
+    src = cms.InputTag("hltInitialStepSeeds"),
+    beamSpot = cms.InputTag("offlineBeamSpot"),
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepSeedTracksLST_cfi.py:6]: is this correct? shouldn't it be `hltOnlineBeamSpot` ?

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltPixelSeedInputLST_cfi.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import FWCore.ParameterSet.Config as cms
+
+hltPixelSeedInputLST = cms.EDProducer('LSTPixelSeedInputProducer',
+    beamSpot = cms.InputTag('offlineBeamSpot'),
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltPixelSeedInputLST_cfi.py:4]: is this correct? shouldn't it be `hltOnlineBeamSpot` ?

File: HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPModulesDevLST_cfi.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import FWCore.ParameterSet.Config as cms
+
+def _addProcessModulesDevLST(process):
+    process.hltESPModulesDevLST = cms.ESProducer('LSTModulesDevESProducer@alpaka',
+        appendToDataLabel = cms.string(''),
+        alpaka = cms.untracked.PSet(
+          backend = cms.untracked.string('')
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPModulesDevLST_cfi.py:7]: can you fix the indentation here?

File: HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHighPtTripletStepSequence_cfi.py [hunk 5]
```diff
@@ -7,3 +7,16 @@
...
 HLTHighPtTripletStepSequence = cms.Sequence(HLTHighPtTripletStepSeedingSequence+hltHighPtTripletStepTrackCandidates+hltHighPtTripletStepTracks+hltHighPtTripletStepTrackCutClassifier+hltHighPtTripletStepTrackSelectionHighPurity)
+
+from Configuration.ProcessModifiers.trackingLST_cff import trackingLST
+trackingLST.toReplaceWith(HLTHighPtTripletStepSequence, HLTHighPtTripletStepSequence.copyAndExclude([HLTHighPtTripletStepSeedingSequence]))
+
+from ..modules.hltHighPtTripletStepTrackCandidatespLSTCLST_cfi import *
+from ..modules.hltHighPtTripletStepTrackspLSTCLST_cfi import *
+from ..modules.hltHighPtTripletStepTrackCutClassifierpLSTCLST_cfi import *
+from ..modules.hltHighPtTripletStepTrackSelectionHighPuritypLSTCLST_cfi import *
+_HLTHighPtTripletStepSequenceLSTSeeding = cms.Sequence(hltHighPtTripletStepTrackCandidatespLSTCLST+hltHighPtTripletStepTrackspLSTCLST+hltHighPtTripletStepTrackCutClassifierpLSTCLST+hltHighPtTripletStepTrackSelectionHighPuritypLSTCLST)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/sequences/HLTHighPtTripletStepSequence_cfi.py:18]: I would slight prefer multi-line breaks here for readability.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltHighPtTripletStepSeedTracksLST_cfi.py (hunk 1 above)
Comment (VourMa) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltHighPtTripletStepSeedTracksLST_cfi.py:6]: Indeed, I will fix it (applies to the other instances of the same comment).

**Comment by mmusich** [comment]
@cmsbuild, please test

File: HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPTTRHBuilderWithoutRefit_cfi.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import FWCore.ParameterSet.Config as cms
+
+def _addProcessTTRHBuilderWithoutRefit(process):
+    process.hltESPTTRHBuilderWithoutRefit = cms.ESProducer("TkTransientTrackingRecHitBuilderESProducer",
+        ComponentName = cms.string('WithoutRefit'),
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPTTRHBuilderWithoutRefit_cfi.py:5]: in the Run-3 HLT menu we generally try to match the name of the producer with the component name, is it possible to adhere to this convenetion also here?

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltGeneralTracks_cfi.py [hunk 7]
```diff
@@ -24,3 +24,58 @@
...
+        pQual = cms.bool(True),
+        tLists = cms.vint32(0)
+    )]
+)
+
+from Configuration.ProcessModifiers.singleIterPatatrack_cff import singleIterPatatrack
+singleIterPatatrack.toReplaceWith(hltGeneralTracks, _hltGeneralTracksSingleIterPatatrack)
+
+_hltGeneralTracksLST = hltGeneralTracks.clone(
+    TrackProducers = ["hltInitialStepTrackSelectionHighPuritypTTCLST", "hltInitialStepTrackSelectionHighPuritypLSTCLST", "hltInitialStepTracksT5TCLST", "hltHighPtTripletStepTrackSelectionHighPurity"],
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltGeneralTracks_cfi.py:43]: as a generic note, I'm proposing to monitor the intermediate track collection produced in the HLT menu at https://github.com/cms-sw/cmssw/pull/46823. 
I suppose we should customize those developments once this PR enters.

**Review by mmusich — COMMENTED** [review]
I tested (successfully) this branch with variations of the HLT phase2 integration tests:
   * `hltPhase2UpgradeIntegrationTests --procModifiers trackingLST`
   * `hltPhase2UpgradeIntegrationTests --procModifiers alpaka,trackingLST`
   * `hltPhase2UpgradeIntegrationTests --procModifiers alpaka,singleIterPatatrack`
   * `hltPhase2UpgradeIntegrationTests --procModifiers alpaka,singleIterPatatrack,trackingLST`

I have only few minor remaining comments, then I think we can go ahead with this PR.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepSeeds_cfi.py [hunk 8]
```diff
@@ -11,5 +11,9 @@
     originRadius = cms.double(0.1),
     useEventsWithNoVertex = cms.bool(True),
     usePV = cms.bool(False),
-    useProtoTrackKinematics = cms.bool(False)
+    useProtoTrackKinematics = cms.bool(False),
+    includeFourthHit = cms.bool(False)
 )
+
+from Configuration.ProcessModifiers.trackingLST_cff import trackingLST
+trackingLST.toModify(hltInitialStepSeeds, includeFourthHit = True)
```
[the remark this answers is not in the document]
Comment (VourMa) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepSeeds_cfi.py:19]: I thought so as well, until I tried it. Bottom line, it only increases the fake+duplicate rate in the non-LST configuration, while only ~preserving the efficiency. You can check the presentation I link in the description (more specifically [here](https://indico.cern.ch/event/1482695/contributions/6257086/attachments/2977298/5241456/LSTInPhase2HLT_20241203.pdf#page=3)) for more details.

[the remark this answers is not in the document]
Comment (mmusich) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepSeeds_cfi.py:19]: Interesting (and to me counterintuitive, albeit goes in the same direction of what was found in https://github.com/cms-sw/cmssw/pull/42820 ) ... thanks for the clarification.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltGeneralTracks_cfi.py (hunk 7 above)
Comment (VourMa) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltGeneralTracks_cfi.py:43]: I agree and, in fact, I have already done this while I was preparing the PR: please check [this commit](https://github.com/SegmentLinking/cmssw/pull/94/commits/8cae22c7f716025c3fa76e936a64637687e458bd). Especially given the presence of https://github.com/cms-sw/cmssw/pull/46823, I thought I would include it in a next PR (after rebasing) but let me know whether it makes sense to include it already and whether you have an comments on the implementation.

File: HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPTTRHBuilderWithoutRefit_cfi.py (hunk 6 above)
Comment (mmusich) [reply HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPTTRHBuilderWithoutRefit_cfi.py:5]: Yes,  thanks

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltGeneralTracks_cfi.py (hunk 7 above)
Comment (mmusich) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltGeneralTracks_cfi.py:43]: > I thought I would include it in a next PR (after rebasing) but let me know whether it makes sense to include it already and whether you have an comments on the implementation.

let's do it as a follow-up step (incidentally, I think we need to change the event content of the step in which the HLT runs in order to make the collections available for the step when the DQM is run).

Comment (VourMa) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltGeneralTracks_cfi.py:43]: Indeed, in my tests I was manually keeping the extra collections in the `cmsDriver` command. Let's follow up with this after this gets merged.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by VourMa** [comment]
@mmusich unless you have any comments on this question of mine:
> I am not sure whether I should use the procModifiers introduced in this PR always with the alpaka procModifier. Feel free to give feedback on that and I can modify the necessary lines.

and modulo a potential squash of commits, I think that, with the latest updates, I should have addressed all of your comments.

**Comment by mmusich** [comment]
>  am not sure whether I should use the procModifiers introduced in this PR always with the alpaka procModifier. Feel free to give feedback on that and I can modify the necessary lines.

as far as I could test the menu runs even without the `alpaka` modifier using only the `trackingLST` modifier. I didn't test `singleIterPatatrack` standalone (I suppose it won't work?) I don't have strong preference. What would be your proposal?

**Comment by VourMa** [comment]
> as far as I could test the menu runs even without the alpaka modifier using only the trackingLST modifier. I didn't test singleIterPatatrack standalone (I suppose it won't work?) I don't have strong preference. What would be your proposal?

In general, I don't particularly like using the `alpaka` procModifier to enable a new algorithm/configuration (as it currently happens with Patatrack for Phase 2 HLT) but that's probably a more general discussion. As long as everything works with the LST modifier, I would keep the PR as is and follow up at a later time to see if we can improve.

**Comment by mmusich** [comment]
>  In general, I don't particularly like using the `alpaka` procModifier to enable a new algorithm/configuration (as it currently happens with Patatrack for Phase 2 HLT) but that's probably a more general discussion

That's the gist of the discussion at https://github.com/cms-sw/cmssw/pull/45508#issuecomment-2239253533, where we decided to take this approach. Or do I misunderstand?

**Comment by VourMa** [comment]
> That's the gist of the discussion at [#45508 (comment)](https://github.com/cms-sw/cmssw/pull/45508#issuecomment-2239253533), where we decided to take this approach. Or do I misunderstand?

My point was having about a specific procModifier (e.g. `patatrack` and not the generic `alpaka`) to enable specific configurations that potentially change the physics. It's not about procModifiers vs. customizations (that's what link in the comment points me to), which has been decided upon and is fine (in my opinion).

**Comment by mmusich** [comment]
> to enable specific configurations that potentially change the physics. 

the point here is that the legacy pixel tracking should not really exist in phase-2 (I think that's beyond discussion). I don't see the point of having a `patatrack` modifier to support a mode that is not envisaged in any possible scenario.

**Comment by mmusich** [comment]
+hlt
   * for https://github.com/SegmentLinking/cmssw/commit/0f32bd093c7eb46986df4065e88dcb285224d4af
   * PR according to description and follow-up review: allows to run the novel LST algorithm in the Phase-2 HLT menu 
   * privately checked modifier-based integration tests, see https://github.com/cms-sw/cmssw/pull/46828#pullrequestreview-2472062523

**Comment by jfernan2** [comment]
+1

**Comment by AdrianoDee** [comment]
+pdmv

**Comment by mmusich** [comment]
@cms-sw/upgrade-l2 can you take a look?

**Comment by Moanwar** [comment]
+Upgrade

**Pull request merged**